### ISSUE TITLE
Windows Sys.rename regression from #12184

### ISSUE THREAD

**Issue by jmid** [ocaml/ocaml#12317]
Unfortunately #12184 introduced a regression of `Sys.rename` under Mingw Windows.
While fixing two corner cases the PR broke a third one, happening when one `Sys.rename`s an empty directory to itself:
```ocaml
Sys.mkdir "_sandbox" 0o755;
Sys.rename "_sandbox" "_sandbox"
```
now fails with `"no such file or directory"` due to the over-eager `RemoveDirectoryW` deletion introduced by #12184.

Possible fixes:
- Revert #12184. This takes us back to the 2 known differences.
- The branch https://github.com/jmid/ocaml/tree/more-sys-rename delays the deletion in case renaming fails.
  This is a step in the right direction, but as @dra27 noticed, this may still leave us between two chairs, if there are insufficient permissions to rename the source directory after having deleted the target.
- David's suggestion was to use the transactional API to make sure we either carry through on the operations or roll back if they fail mid-way https://learn.microsoft.com/en-us/windows/win32/fileio/transactional-ntfs-portal
- ...

I'm opening the issue to document the regression until it is resolved (other suggestions are also welcome).

**Comment by avsm** [comment]
You could also just exit immediately from the function with success if the two input strings are the same.  There's no need to actually perform an identity filesystem operation is there?

**Comment by nojb** [comment]
> You could also just exit immediately from the function with success if the two input strings are the same. There's no need to actually perform an identity filesystem operation is there?

I may be missing something, but isn't it the case that this would not be enough since different pathnames can refer to the same directory? (Think paths containing `.`, `..`, different capitalization, forward/backwards slashes, etc).

**Comment by jmid** [comment]
> I may be missing something, but isn't it the case that this would not be enough since different pathnames can refer to the same directory? (Think paths containing ., .., different capitalization, forward/backwards slashes, etc).

Indeed, I think some form of normalization is required in general (also to handle a full and a relative path to the same dir).

**Comment by xavierleroy** [comment]
Allow me to step back a little.  Do we really want to guarantee that `Sys.rename` behaves as specified in the latest POSIX standards even under Windows?  Emulating POSIX under Windows is the job of Cygwin, but is it the job of the OCaml runtime system?  How much effort @dra27 will spend chasing this goal?

Instead, could we just document the common cases where `Sys.rename` works as expected both under POSIX and under Windows, and leave the other cases explicitly unspecified?

**Comment by xavierleroy** [comment]
Here is how `rename` is specified in ISO C 2011: "The rename function causes the file whose name is the string pointed to by old to be henceforth known by the name given by the string pointed to by new. The file named old is no longer accessible by that name. If a file named by the string pointed to by new exists prior to the call to the rename function, the behavior is implementation-defined.".

**Comment by alainfrisch** [comment]
> The file named old is no longer accessible by that name.

A literal reading of that implies that renaming a file to itself makes it inaccessible?

**Comment by xavierleroy** [comment]
> A literal reading of that implies that renaming a file to itself makes it inaccessible?

At the same time it falls in the "behavior is implementation-defined" part.

We can write a different spec.  My point is that it's hopeless to expect POSIX behavior from `Sys.rename` even under Windows, so some amount of underspecification is necessary.

**Comment by dra27** [comment]
There's a slight bit of under-reporting here - the transactional suggestion (offline) was only to do with the atomicity of the operation, and was a suggestion to which I added "But I think this might be taking it too far 🙂". We have a transactional file system, but I'm happy to be afraid to use it!

Given that you can rename a file to itself in the shell, too, so the regressed behaviour is certainly surprising (`ren foo foo` / `Rename-Item foo foo` are both perfectly fine). If @jmid's branch is presently correct (ignoring atomicity), I'd suggest we go with that (for now, and probably forever).

**Comment by dra27** [comment]
In terms of worrying about these differences, it is (and has been) painful when they are subtly found in other systems (both Dune and opam, for example, have experienced various bits of pain in the past with things like this). That doesn't have to be our problem in ocaml/ocaml, but for me at least it's a worthy goal to have the system functions in the standard library behave as far as we can in as consistent a way as possible between the two platforms.

**Comment by jmid** [comment]
I realized today that #12184 has been included on the candidate 5.1 branch.
I've therefore opened PR #12320 based on the mentioned branch.

**Comment by jmid** [comment]
Solved with merge of #12320, thus closing.
